colossalai/pipeline/schedule/zero_bubble_pp.py

```diff
@@ -0,0 +1,819 @@
+            grad=output_obj_grad,
+            inputs=input_obj,
+            retain_graph=True,
+        )
+        return input_obj.grad
+
+    def backward_w_step(
+        self,
+        model_chunk: Union[ModuleList, Module],
+        model_chunk_id: int,
+        optimizer: OptimizerWrapper,
+        output_obj: Union[dict, torch.Tensor],
+        output_obj_grad: Optional[dict],
+    ):
+        """Backward dw step of the pipeline; we calculate "dw = x*dy" here;
+
+        Args:
+            model_chunk (ModuleList or Module): Model Chunk to be run;
+            model_chunk_id (int): The current model chunk idx;
+            optimizer (OptimizerWrapper): Optimizer to update the model
+            output_obj (Union[dict, torch.Tensor]): y.
+            output_obj_grad (dict): dy.
+
+        Returns:
+            Nothing need to return; we only calculate dw then update w;
+        """
+        # calculate bwd w step ; only dw = x*dy;
+
+        if model_chunk_id == 1 and self.stage_manager.is_first_stage(ignore_chunk=True):
+            # loss backward; output_obj is loss
+            output_obj_grad = None
+        optimizer.backward_by_grad(
+            tensor=output_obj,
+            grad=output_obj_grad,
+            inputs=list(model_chunk[model_chunk_id].parameters()),
+            retain_graph=False,
+        )
+
+    def schedule_f(
+        self,
+        scheduled_node,
+        model_chunk: torch.nn.ModuleList,
+        model_chunk_id: int,
+        criterion: Callable,
+        accum_loss: Optional[torch.Tensor] = None,
+        outputs: Optional[List[Any]] = None,
+    ):
+        """A complete forward schedule; Include recv fwd --> cal fwd --> send fwd;
+
+        Args:
+            scheduled_node:
+            model_chunk (ModuleList or Module): Model Chunk to be run;
+            model_chunk_id (int): The current model chunk idx;
+            criterion (Callable): loss function;
+            accum_loss (Optional[torch.Tensor], optional): Accumulated loss. Defaults to None.
+            outputs (Optional[List[Any]], optional): List to store the output of the last stage (final output). Defaults to None.
+
+        Returns:
+            Nothing.
+        """
+        micro_batch = self.load_micro_batch(model_chunk_id=model_chunk_id)
+        # Step1: recv fwd
+        if model_chunk_id == 0:
+            # is first stage; get input from func param
+            if self.stage_manager.is_first_stage(ignore_chunk=True):
+                input_obj = micro_batch
+            else:
+                input_obj = self.recv_forward_buffer[model_chunk_id].pop(0)
+        else:
+            # is last stage; recv from local
+            if self.stage_manager.is_last_stage(ignore_chunk=True):
+                input_obj = self.local_send_forward_buffer.pop(0)
+            # not last stage; recv from next
+            else:
+                input_obj = self.recv_forward_buffer[model_chunk_id].pop(0)
+        input_obj.requires_grad_()
```

Comment:
use tree_map() like line 454

Reply:
> use tree_map() like line 454

Replace with treemap in https://github.com/hpcaitech/ColossalAI/pull/6034/commits/ce58d8e8bf8c8807eb37b29fff8495b155279274;
Now, like "tree_map(torch.Tensor.requires_grad_, input_obj)"